### ISSUE TITLE
Scanned barcode shoud set the card barcode, not the ID when they differ

### ISSUE THREAD

**Issue by mvglasow** [CatimaLoyalty/Android#1453]
By default, Catima assumes card ID and barcode to be the same. Scan the barcode, Catima will detect its format, read the code and store the barcode and card ID. So far, so good.

When the barcode and card ID differ, the workflow is as follows (as of version 2.25.1):

* Scan the barcode, or enter the ID manually and choose a barcode format. Your input will be used as the ID and barcode value.
* In the drop-down list for the barcode value, choose to set the barcode value manually.

Not sure how I’m expected proceed from there. Personally, I scan the barcode first, copy the ID, select to enter the barcode manually, then paste the scanned and copied ID into the barcode, then change the ID – but I find this workflow less than perfect. I would expect to scan the barcode and enter the ID manually (if it differs) – not the other way round.

Suggested improvement:

* Scanning a barcode (or entering it manually) will always set the barcode.
* When creating a new card, or changing the barcode for a card where the ID matches the barcode, scanning the barcode will also set the ID.
* After this step, the user can choose to manually enter an ID which differs from the barcode. This could be accomplished by changing the ID field into a drop-down one (similar to the barcode field now), where the user can select “Same as barcode” or “Enter manually”. The barcode field would then be a simple text field, disabled when the barcode type is set to “No barcode”.

This would save me a few steps when importing a new card with an ID which differs from the QR code.

**Comment by TheLastProject** [comment]
Yeah, looking at it more closely I agree the flow is the wrong way around and all actions should by default affect "Barcode value", not "Card ID".

I fear this might need a database migration to properly fix, but your idea and suggested flow sounds correct to me.

**Comment by TheLastProject** [comment]
(Tagging with "severity: major" as it can make setting a custom card ID very unfriendly and "common: occasional" as most users don't care but many stores do use a different ID than value)

**Comment by obfusk** [comment]
> Yeah, looking at it more closely I agree the flow is the wrong way around and all actions should by default affect "Barcode value", not "Card ID".

Yeah. I think I mentioned this before. I can probably think of a fix that doesn't require a migration. But the latter is probably cleaner.

**Comment by obfusk** [comment]
> That could be worked around in code

That's what I suggested yes :)

> Still somewhat complex, but needs to be done only once and only if there is legacy data

We'd also need to deal with the changed format in the importer. It's cleaner but a lot more work. Not sure if it's worth it. Up to @TheLastProject to decide :)

**Comment by TheLastProject** [comment]
Sorry for the extremely delayed response, life got hectic and responding to this ended up somewhere on my todo list.

While I generally dislike doing database migrations and prefer to avoid them whenever possible (simply because making a single mistake there can completely destroy user data without any way to recover it) I do feel a database change is the cleaner solution here and probably the route we should take. A migration like this can be tested and will long-term cause less complexity than working around things in code, even if it's more pain at first.

Switching the barcode and card ID around would indeed also affect the importer (as old imports may have the barcode id "in the wrong field"), affect the exporter (we'd need an export format version bump) and affect Gadgetbridge integration (although it would be fairly minor and easy to communicate with them, just bump the major version and tell them to read barcodeid instead of cardid on the new major version of the content provider). All in all it would be a group of minor breaking changes, nothing too scary, just would need thorough testing and good communication with the Gadgetbridge devs.
